transport/peer/hostport/peer_test.go

```diff
@@ -0,0 +1,179 @@
+		appliedFunc        func(*Peer)
+		expectedIdentifier string
+		expectedHostPort   string
+		expectedStatus     transport.PeerStatus
+		expectedAgent      transport.Agent
+	}
+	tests := []testStruct{
+		func() (s testStruct) {
+			s.msg = "create"
+			s.pi = PeerIdentifier("localhost:12345")
+			s.agent = transporttest.NewMockAgent(mockCtrl)
+
+			s.appliedFunc = func(p *Peer) {}
+
+			s.expectedIdentifier = "localhost:12345"
+			s.expectedHostPort = "localhost:12345"
+			s.expectedAgent = s.agent
+			s.expectedStatus = transport.PeerStatus{
+				PendingRequestCount: 0,
+				ConnectionStatus:    transport.PeerUnavailable,
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "start request"
+			agent := transporttest.NewMockAgent(mockCtrl)
+			agent.EXPECT().NotifyStatusChanged(gomock.Any()).Times(1)
+			s.agent = agent
+
+			s.appliedFunc = func(p *Peer) {
+				p.StartRequest()
+			}
+
+			s.expectedAgent = s.agent
+			s.expectedStatus = transport.PeerStatus{
+				PendingRequestCount: 1,
+				ConnectionStatus:    transport.PeerUnavailable,
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "start request stop request"
+			agent := transporttest.NewMockAgent(mockCtrl)
+			agent.EXPECT().NotifyStatusChanged(gomock.Any()).Times(2)
+			s.agent = agent
+
+			s.appliedFunc = func(p *Peer) {
+				done := p.StartRequest()
+				done()
+			}
+
+			s.expectedAgent = s.agent
+			s.expectedStatus = transport.PeerStatus{
+				PendingRequestCount: 0,
+				ConnectionStatus:    transport.PeerUnavailable,
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "start 5 stop 2"
+			agent := transporttest.NewMockAgent(mockCtrl)
+			agent.EXPECT().NotifyStatusChanged(gomock.Any()).Times(7)
+			s.agent = agent
+
+			s.appliedFunc = func(p *Peer) {
+				done1 := p.StartRequest()
+				p.StartRequest()
+				p.StartRequest()
+				done2 := p.StartRequest()
+				done1()
+				p.StartRequest()
+				done2()
+			}
+
+			s.expectedAgent = s.agent
+			s.expectedStatus = transport.PeerStatus{
+				PendingRequestCount: 3,
+				ConnectionStatus:    transport.PeerUnavailable,
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "start 5 stop 5"
+			agent := transporttest.NewMockAgent(mockCtrl)
+			agent.EXPECT().NotifyStatusChanged(gomock.Any()).Times(10)
+			s.agent = agent
+
+			s.appliedFunc = func(p *Peer) {
+				for i := 0; i < 5; i++ {
+					done := p.StartRequest()
+					defer done()
+				}
+			}
+
+			s.expectedAgent = s.agent
+			s.expectedStatus = transport.PeerStatus{
+				PendingRequestCount: 0,
+				ConnectionStatus:    transport.PeerUnavailable,
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "set status"
+
+			s.appliedFunc = func(p *Peer) {
+				p.SetStatus(transport.PeerAvailable)
+			}
+
+			s.expectedStatus = transport.PeerStatus{
+				PendingRequestCount: 0,
+				ConnectionStatus:    transport.PeerAvailable,
+			}
+			return
+		}(),
+	}
+
+	for _, tt := range tests {
+		if tt.pi == PeerIdentifier("") {
```

Comment:
just as an fyi since it looks like you're going to modify this test, we should use subtests with our table tests for easier testing: https://blog.golang.org/subtests

```
for _, tt := range tests {
 		t.Run(tt.msg, func(t *testing.T){
 		 		if tt.pi == PeerIdentifier("") {
...
```